### PR TITLE
Reduce cpu and memory alloc around configuration

### PR DESCRIPTION
**Pull request by oskar-szwajkowski**
Use short-circuit when there is no key in methodMap
Pull up metadataCache to be static field in ConfigurationFactory

I could see this static cache being behind some ENV variable, but its static memory footprint should be below 1MB per class-loader (tested with galaxy-trino object store connector)

### CODE REVIEW

File: configuration/src/main/java/io/airlift/configuration/ConfigurationFactory.java [hunk 1]
```diff
@@ -137,7 +138,7 @@ public ConfigurationFactory(Map<String, String> properties, WarningsMonitor warn
     {
         this.properties = ImmutableMap.copyOf(properties);
         this.warningsMonitor = warningsMonitor;
-        this.monitor = monitor;
+        ConfigurationFactory.monitor = monitor;
```
Comment (electrum) [inline configuration/src/main/java/io/airlift/configuration/ConfigurationFactory.java:141]: We shouldn't set a static field from a constructor. It's confusing and not thread safe.

Instead, we should refactor the tests so that `Problems.Monitor` is not needed. It appears that the tests could access `Problems` directly, but you'll need to try refactoring it and see what it looks like.

**Review by electrum — CHANGES_REQUESTED** [review]

File: configuration/src/main/java/io/airlift/configuration/ConfigurationFactory.java (hunk 1 above)
Comment (oskar-szwajkowski) [reply configuration/src/main/java/io/airlift/configuration/ConfigurationFactory.java]: Changed it so that it's not static anymore.
I did try to remove monitor but it require more code changes that I wanted to fit in this PR

**Review by wendigo — APPROVED** [review]

**Comment by oskar-szwajkowski** [comment]
@electrum Can you review it?

File: configuration/src/main/java/io/airlift/configuration/ConfigurationFactory.java [hunk 2]
```diff
@@ -84,6 +84,15 @@ public class ConfigurationFactory
     private static final Validator VALIDATOR;
 
     private static final Splitter VALUE_SPLITTER = Splitter.on(",").omitEmptyStrings().trimResults();
+    private static final LoadingCache<Class<?>, ConfigurationMetadata<?>> metadataCache = CacheBuilder.newBuilder()
```
Comment (electrum) [inline configuration/src/main/java/io/airlift/configuration/ConfigurationFactory.java:87]: Name this `METADATA_CACHE` as it's a constant

File: configuration/src/main/java/io/airlift/configuration/ConfigurationFactory.java [hunk 3]
```diff
@@ -84,6 +84,15 @@ public class ConfigurationFactory
...
+            .build(new CacheLoader<>()
```
Comment (electrum) [inline configuration/src/main/java/io/airlift/configuration/ConfigurationFactory.java:88]: This can simplify to
```java
.build(CacheLoader.from(ConfigurationMetadata::getConfigurationMetadata));

File: configuration/src/main/java/io/airlift/configuration/ConfigurationFactory.java [hunk 4]
```diff
@@ -286,6 +287,13 @@ <T> void registerConfigDefaults(ConfigDefaultsHolder<T> holder)
         registeredDefaultConfigs.put(holder.getConfigKey(), holder);
     }
 
+    @VisibleForTesting
+    static void cleanMetadataCache()
```
Comment (electrum) [inline configuration/src/main/java/io/airlift/configuration/ConfigurationFactory.java:291]: Do we need this if we get rid of `Problems.Monitor`?

**Comment by wendigo** [comment]
@oskar-szwajkowski can you fix failures?

File: configuration/src/test/java/io/airlift/configuration/TestConfigurationFactory.java [hunk 5]
```diff
@@ -174,12 +168,11 @@ public void testConfigurationThroughDeprecatedConfig()
...
         properties.put("string-b", "this is b");
-        TestMonitor monitor = new TestMonitor();
-        Injector injector = createInjector(properties, monitor, binder -> configBinder(binder).bindConfig(DeprecatedConfigPresent.class));
+        TestWarningsMonitor warningsMonitor = new TestWarningsMonitor();
+        Injector injector = createInjector(properties, binder -> configBinder(binder).bindConfig(DeprecatedConfigPresent.class), warningsMonitor);
         DeprecatedConfigPresent deprecatedConfigPresent = injector.getInstance(DeprecatedConfigPresent.class);
-        monitor.assertNumberOfErrors(0);
-        monitor.assertNumberOfWarnings(1);
-        monitor.assertMatchingWarningRecorded("Configuration property 'string-a' is deprecated and should not be used");
+        assertThat(warningsMonitor.messages).hasSize(1);
```
Comment (electrum) [inline configuration/src/test/java/io/airlift/configuration/TestConfigurationFactory.java:174]: Add a `messages()` getter

File: configuration/src/main/java/io/airlift/configuration/ConfigurationMetadata.java [hunk 6]
```diff
@@ -72,13 +60,13 @@ static <T> ConfigurationMetadata<T> getConfigurationMetadata(Class<T> configClas
...
 
-    private ConfigurationMetadata(Class<T> configClass, Monitor monitor)
+    private ConfigurationMetadata(Class<T> configClass)
     {
         if (configClass == null) {
             throw new NullPointerException("configClass is null");
         }
 
-        this.problems = new Problems(monitor);
+        this.problems = new Problems();
```
Comment (electrum) [inline configuration/src/main/java/io/airlift/configuration/ConfigurationMetadata.java:69]: Move this initialization to the field declaration

**Review by electrum — APPROVED** [review]

**Comment by oskar-szwajkowski** [comment]
> @oskar-szwajkowski can you fix failures?

Done, can you merge & release?

File: configuration/src/test/java/io/airlift/configuration/TestMonitor.java [hunk 7]
```diff
@@ -18,17 +18,18 @@
...
+import java.util.LinkedHashSet;
+import java.util.Set;
 
 import static org.assertj.core.api.Assertions.assertThat;
 import static org.assertj.core.api.Assertions.fail;
 
 class TestMonitor
         implements Problems.Monitor
 {
-    private final List<Message> errors = new ArrayList<>();
```
Comment (wendigo) [inline configuration/src/test/java/io/airlift/configuration/TestMonitor.java:30]: Why? Seems unrelated. Changes order of error messages.

Comment (oskar-szwajkowski) [reply configuration/src/test/java/io/airlift/configuration/TestMonitor.java:30]: this class gets deleted in later commit, but when I've added changes to Problems in commit no. 2, then it sometimes contained duplicate messages

File: configuration/src/test/java/io/airlift/configuration/TestConfigurationFactory.java [hunk 8]
```diff
@@ -451,16 +436,16 @@ public void testUsedPropertiesWithFailure()
...
 
-        ConfigurationFactory configurationFactory = new ConfigurationFactory(properties, null, new TestMonitor());
+        ConfigurationFactory configurationFactory = new ConfigurationFactory(properties);
         configurationFactory.registerConfigurationClasses(ImmutableList.of(binder -> configBinder(binder).bindConfig(AnnotatedSetter.class)));
         List<Message> messages = configurationFactory.validateRegisteredConfigurationProvider();
         assertMessagesMatch(messages, ImmutableList.of(".*Invalid value 'invalid' for type boolean \\(property 'boolean-value'\\).*AnnotatedSetter.*"));
         assertThat(configurationFactory.getUsedProperties()).isEqualTo(properties.keySet());
     }
 
-    private static Injector createInjector(Map<String, String> properties, TestMonitor monitor, Module module)
```
Comment (wendigo) [inline configuration/src/test/java/io/airlift/configuration/TestConfigurationFactory.java:461]: do we need to switch order here? Makes diff bigger than it could be

**Review by wendigo — APPROVED** [review]
